java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,99 @@
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+import arcs.core.data.Claim
+import arcs.core.data.expression.Expression
+
+/** Field [Identifier]. */
+typealias Identifier = String
+
+/** Lists of [Identifier]s imply an [AccessPath]. */
+typealias Path = List<Identifier>
+
+/**
+ * [Deduction]s represent how inputs of a Paxel [Expression] contribute to its outputs.
+ *
+ * [Deduction]s recursively associate identifiers to field paths. These relationships can be
+ * directly translated into [Claim]s and [AccessPath]s.
+ */
+sealed class Deduction {
+
+  /** Substitute aliases in all paths. */
+  abstract fun substitute(aliases: Scope): Deduction
+
+  /** Union of two [Deduction] collections. */
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** A representation of a lookup claim in a Paxel [Expression]. */
+  data class Equal(val path: Path = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Return [Equal] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Equal = Equal(
+      path[0].let { identifier ->
+        val association = aliases.associations.getOrDefault(identifier, Equal(identifier))
+        if (association is Equal) association.path else throw UnsupportedOperationException(
+          "Cannot substitute ${association::class.simpleName} for an Identifier."
+        )
+      } + path.drop(1)
+    )
+
+    /** Union of a [Equal] and a [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> Derive(this.path, other.path)
+      is Derive -> Derive(setOf(this.path) + other.paths)
+      is Scope -> throw UnsupportedOperationException(
+        "Union of Equal and Scope is not well defined."
+      )
+    }
+  }
+
+  /** A representation of a Derivation claim in a Paxel [Expression]. */
+  data class Derive(val paths: Set<Path> = emptySet()) : Deduction() {
+
+    constructor(vararg paths: Path) : this(paths.toSet())
+
+    /** Union of a [Derive] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> Derive(this.paths + setOf(other.path))
+      is Derive -> Derive(paths + other.paths)
+      is Scope -> throw UnsupportedOperationException(
+        "Union of Derive and Scope is not well defined."
+      )
+    }
+
+    /** Substitute all aliases as a new [Derive] object. */
+    override fun substitute(aliases: Scope) = Derive(
+      paths.map { path -> Equal(path).substitute(aliases).path }.toSet()
+    )
+  }
+
+  /** Associates an [Identifier] with a group of [Deduction]s. */
+  data class Scope(
```

Comment:
`Scope` is not a great name. Perhaps, `Context` or even `IdentifierDeductions`? (We can bike shed as a final step.)

Reply:
I think this is trying to be parallel to nomenclature in Paxel ExpressionEvaluaor. Paxel expressions do use scopes. All `QualifiedExpression` operations (from, where, select, let, etc) are from `Sequence<Scope>` to `Sequence<Scope>`.  Scopes in paxel are immutable, and chained. e.g.

```
val newScope = scope.builder().set("x", 1).build()
val newScope2 = scope.builder().set("y", 2).build()
val z = newScope2.lookup("x") // == 1
```

Scopes have parent scopes and are chained. So for example, various Paxel operations that "mutate" scopes iteratively, in actuality, create new scopes, with temporary variables, which are consumed by latter operations, but go out of existence when the next item is iterated. 

Something like `from x in foo where x < 10 select Foo { z: x + 1 }` is really just `foo.map { scope.builder().set("x", it).build() }.filter { it.lookup("x") < 10 }.map { newScope("z" to it.lookup("x") + 1) }` (pseudo code)

Paxel is just a map-reduce-filter chain over scopes.